### DISCUSSION TITLE
Pmetrics can't find Gfortran

### DISCUSSION THREAD

**Discussion by CeciKtz** [LAPKB/Pmetrics#224] [General]
Hi! when I run PMbuild() it can´t find Fortran, I tried several times to download it automatically or manually. I have checked the manual download goes to the default location, checked its installation but Pmetrics won't find gfortran when it is needed
When I run PMbuild() a get this: 
> PMbuild()
Pmetrics cannot find required compiled binary files.

Pmetrics cannot detect gfortran and will attempt to download and install all components.
Do you agree? (Y/N)y
Checking for Brew 
Pmetrics did not find the Homebrew Package Manager and will attempt to download and install it.
Do you wish to proceed? (Y/N)y
A new terminal will open and ask for permission to install.  Enter your computer password.
Press any key when Homebrew installation is complete.

The terminal shows this message: 
~ % /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library/Pmetrics/mac/install_homebrew.sh ; exit;
curl: (22) The requested URL returned error: 404

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

Then R console goes with:

gcc 13.2.0
Pmetrics found GCC.
==> Auto-updating Homebrew...
Adjust how often this is run with HOMEBREW_AUTO_UPDATE_SECS or disable with
HOMEBREW_NO_AUTO_UPDATE. Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
==> Homebrew collects anonymous analytics.
Read the analytics documentation (and how to opt-out) here:
  https://docs.brew.sh/Analytics
No analytics have been recorded yet (nor will be during this `brew` run).

==> Homebrew is run entirely by unpaid volunteers. Please consider donating:
  https://github.com/Homebrew/brew#donations

Installing from the API is now the default behaviour!
You can save space and time by running:
  brew untap homebrew/core
==> Auto-updated Homebrew!
Updated 2 taps (homebrew/core and homebrew/cask).
==> New Formulae
chainhook
==> New Casks
virtualdj

ERROR: Pmetrics did not install gfortan automatically.
Please install gfortran manually and then run PMbuild().
Go to http://www.lapk.org/Pmetrics_install.php for help.
ERROR: Pmetrics did not install gfortran automatically.
Please install gfortran manually and then run PMbuild().
Go to http://www.lapk.org/Pmetrics_install.php for help.

Did anybody experience this problem? My IOs was recently updated to Sonoma 14.2 and after that I can't run Pmetrics, I think that maybe solving the automatic download  would make it run properly, I hope so. Any help is a lot. Thanks!!

**Comment by CeciKtz** [comment]
I could make Pmetrics find fortran and compile. But a new error appears when it start to do en NPrun: starts to compile again (like I was called 'PMbuild()'. The iteration stops and the terminal shows this message:

WARNING: Operating system not declared.
 WARNING: Can not initialize system calls.
 
 Possible command line error, try:
 C:\> npbig.exe <DOS,Win,XP,Vista,or Win7>
 unix$ npbig <MacOSX,Darwin,BSD>
 
 NPAG Exiting with status 2

Does anyone know how to install manually npbig? Thanks!!

**Comment by Siel** [comment]
That error you are seeing is part of the fortran code checks for some of the configuration files needed to run NPAG:


<img width="599" alt="image" src="https://github.com/LAPKB/Pmetrics/assets/1023006/f3e9cdd5-743f-423e-820d-a3045f794718">


My guess is that your code is getting compiled but there is something wrong with you Pmetrics installation. Can I have the full log?

**Reply by CeciKtz** [reply]
Yes! what should I do to give you full log?

**Reply by CeciKtz** [reply]
This is in a terminal window called np_prep:

Last login: Fri May 10 11:36:12 on ttys002
mariaceciliakravetz@MacBook-Pro-de-Maria-2 ~ % /Users/mariaceciliakravetz/Documents/Pmetrics/Examples/Runs/1/etc/np_prep ; exit;


 IN THIS "PC" PROGRAM, BECAUSE OF PC SPACE LIMITATIONS,
 THE NUMBER OF POPULATION SUBJECTS IS CURRENTLY LIMITED TO  800,

 THE NUMBER OF OBSERVATIONS PER SUBJECT IS LIMITED TO  150,

 THE NUMBER OF RANDOM PARAMETERS IS LIMITED TO 30,

 THE NUMBER OF ACTIVE GRID POINTS IN THE JOINT DENSITY RESULTS
 	   FILE IS LIMITED TO       1500,

 THE NUMBER OF OUTPUT EQUATIONS IS LIMITED TO  7.


 WARNING: Operating system not declared.
 WARNING: Can not initialize system calls.
 
 Possible command line error, try:
 C:\> npbig.exe <DOS,Win,XP,Vista,or Win7>
 unix$ npbig <MacOSX,Darwin,BSD>
 
 NPAG Exiting with status 2
 

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Proceso completado]

There's another one 'np_run' with this content:

Last login: Fri May 10 17:57:35 on ttys002
mariaceciliakravetz@MacBook-Pro-de-Maria-2 ~ % /Users/mariaceciliakravetz/Documents/Pmetrics/Examples/Runs/1/etc/np_run ; exit;
At line 12710 of file NPeng_120.f (unit = 25)
Fortran runtime error: Cannot open file 'extnum': No such file or directory

Error termination. Backtrace:
#0  0x15ffeed2f
#1  0x15ffef8d7
#2  0x15fff01a7
#3  0x1600df0a3
#4  0x1600df5bf
#5  0x104af7d5b
#6  0x104b03b9f
#7  0x104b0473f

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Proceso completado]

And a last one called 'npscript with this:

Last login: Fri May 10 17:58:54 on ttys001
mariaceciliakravetz@MacBook-Pro-de-Maria-2 ~ % /Users/mariaceciliakravetz/Documents/Pmetrics/Examples/Runs/1/etc/npscript ; exit;
/Users/mariaceciliakravetz/Documents/Pmetrics/Examples/Runs/1/etc/npscript: line 4: PMcontrol: No such file or directory
f951: Fatal Error: Cannot open file 'npagdriv.f'
compilation terminated.
/Users/mariaceciliakravetz/Documents/Pmetrics/Examples/Runs/1/etc/npscript: line 8: ./np_run: No such file or directory

Cleaning up....

mkdir: inputs: File exists
mkdir: outputs: File exists
mkdir: wrkcopy: File exists
mkdir: etc: File exists
mv: instr.inx: No such file or directory
mv: log.txt: No such file or directory
mv: PMcontrol: No such file or directory
mv: model.for: No such file or directory
mv: model.txt: No such file or directory
mv: rename XQZPJ*.ZMQ to wrkcopy/XQZPJ*.ZMQ: No such file or directory
mv: rename npag*.* to etc/npag*.*: No such file or directory
rm: CHMAX*.*: No such file or directory
rm: fort.*: No such file or directory
mv: rename np_prep* to etc/np_prep*: No such file or directory
mv: rename np_run* to etc/np_run*: No such file or directory
mv: ex.csv: No such file or directory
Loading required package: Pmetrics

CRITICAL: Execute PMbuild() in R to complete Pmetrics installation.

Generating report based on the plotly template...
Error in loadNamespace(x) : there is no package called ‘pandoc’
Calls: <Anonymous> ... loadNamespace -> withRestarts -> withOneRestart -> doWithOneRestart
Execution halted
Error in pander::openFileInOS("/Users/mariaceciliakravetz/Documents/Pmetrics/Examples/Runs/1/outputs/NPAGreport.html") : 
  File not found!
Execution halted
mv: npscript: No such file or directory

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Proceso completado]

I don't know if these is helpful for understanding the problem. Thanks again.

**Reply by Siel** [reply]
I have seen the second error before there should be a file called exnum inside the run folder with a number inside (Pmetrics should create that file by default). The first error seems to be caused because the program cannot detect your OS. I do not know how NPAG get that information maybe another file? @masyamada @mnneely ?

Your system seems to be having really serious issues with fortran, which one is Pmetrics using? do you have a different machine you can test to install the package?

**Reply by mnneely** [reply]
Sorry to jump in the middle. Pmetrics supplies the OS as a command line argument when running np_prep. That’s why there was an error trying to run the file directly from the terminal.

**Comment by Siel** [comment]
Please run PMbuild() and paste here the whole output.

Thanks!

**Comment by CeciKtz** [comment]
Wien I run PMbuild() R shows:

> PMbuild()
Pmetrics cannot find required compiled binary files.
Pmetrics has detected gfortran and will compile required binary files.
If the registration time outs, please re-run PMbuild(skipRegistration = TRUE) Pmetrics has anonymously registered your installation of this version.
LAPKB does not collect or store any personal or identifying information.

Compiling 1 of 9: NPprep...
Compiling 2 of 9: NPeng...
Compiling 3 of 9: ITprep...
Compiling 4 of 9: ITeng...
Compiling 5 of 9: ITerr...
Compiling 6 of 9: SIMeng...
Compiling 7 of 9: DOprep...
Compiling 8 of 9: DOeng...
Compiling 9 of 9: mb2csv...
All packages installed and permanent Fortran modules compiled.
> 

When I caller NPrun it starts the same way

**Reply by mnneely** [reply]
This seems to suggest that the program was able to compile and Fortran is okay.

> When I caller NPrun it starts the same way

Do you mean that it again says it can’t find Fortran, or does it say “WARNING: Operating system not declared.
WARNING: Can not initialize system calls.”?

What output do you get with this command: `Pmetrics:::getOS()`?

**Reply by CeciKtz** [reply]
When I run NPrun I get in R the same when I call PMbuild():

Model solver mode: Algebraic
Number of compartments: 2, including an absorptive compartment
Primary Variables: Ka, Ke, V, Tlag1
Covariates in data file:  wt, africa, age, gender, height
Covariates used in model file:  WT, AFRICA, AGE, GENDER, HEIGHT
Secondary Variables: KCP, KPC
Model conditions: no bioavailability term defined, initial conditions are zero, lag term defined
Number of cycles to run: 100

Pmetrics cannot find required compiled binary files.
Pmetrics has detected gfortran and will compile required binary files.
If the registration time outs, please re-run PMbuild(skipRegistration = TRUE) Pmetrics has anonymously registered your installation of this version.
LAPKB does not collect or store any personal or identifying information.

Compiling 1 of 9: NPprep...
Compiling 2 of 9: NPeng...
Compiling 3 of 9: ITprep...
Compiling 4 of 9: ITeng...
Compiling 5 of 9: ITerr...
Compiling 6 of 9: SIMeng...
Compiling 7 of 9: DOprep...
Compiling 8 of 9: DOeng...
Compiling 9 of 9: mb2csv...
All packages installed and permanent Fortran modules compiled.
[1] "/Users/mariaceciliakravetz/Documents/Pmetrics/Examples/Runs/1/outputs"
---
When I open the ´outputs´ folder in Run 1, I find three files which are terminal windows with the information I showed before.
---
Something quite weird happened.
I called Pmetrics:::getOS() and I got:

> Pmetrics:::getOS()
[1] 1

After that, everytime I try to write in the console I get an error:
Error in fetch(key) : 
lazy-load database '/Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library/Pmetrics/help/Pmetrics.rdb' is corrupt

**Reply by mnneely** [reply]
Well, Pmetrics is detecting your OS okay, and you are able to compile the Fortran modules, but somehow it isn't registering with Pmetrics that you have compiled. Which version of Pmetrics are you running? It will tell you if you look at the list of packages in the lower right window of Rstudio.

**Reply by CeciKtz** [reply]
Yes! Pmetrics version is 2.1.1

**Comment by mnneely** [comment]
OK, I've discovered a recent change in R that might have broken something.

Tell me the output of the following two commands:

`length(grep("32-bit", utils::sessionInfo()))`
and
`Pmetrics:::getBits()`

**Comment by CeciKtz** [comment]
Of course!
 >length(grep("32-bit", utils::sessionInfo()))
[1] 0

> Pmetrics:::getBits()
[1] 32

it is ok?

**Reply by mnneely** [reply]
I think this might be the problem or at least part of it. Pmetrics detects whether you are running a 32- or 64-bit system using that `sessionInfo()` function in R. That function has apparently changed. Because the response to the first command was "0", that means it did not detect the phrase "32-bit", which means you have a 64-bit computer, as is the case for all newer computers. The second command should have returned "64". 

I have updated the internal Pmetrics function getBits in the latest version in the dev branch. If you can install that, it might fix your problems.

`remotes::install_github("LAPKB/Pmetrics@dev")`

**Reply by CeciKtz** [reply]
Well, I re-installed Pmetrics, when running the installation an error appeared:

installing to /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/library/00LOCK-Pmetrics/00new/Pmetrics/libs
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** testing if installed package can be loaded from temporary location
**Failed with error:  ‘there is no package called ‘PmetricsData’’**
** checking absolute paths in shared objects and dynamic libraries
** testing if installed package can be loaded from final location
**Failed with error:  ‘there is no package called ‘PmetricsData’’**
** testing if installed package keeps a record of temporary installation path
* DONE (Pmetrics)

Then I called PMbuild() which seems to run normally

And when calling PMtest() I get:
> PMtest()
Error in find.package(package, lib.loc, verbose = verbose) : 
  there is no package called ‘PmetricsData’

Thank you so much, Michael and team, for the tremendous support

**Comment by mnneely** [comment]
Oh, sorry for the trouble with this. We’re in the process of trying to get the package onto CRAN, and clearly have a few bugs. You should be able to use `remotes::install_github(“LAPKB/PmetricsData”)`. We had to split off the examples and data for the package into a separate one to keep the Pmetrics package size closer to CRAN guidelines.

**Reply by CeciKtz** [reply]
Thanks for the support!
Yesterday I finished installation which looked OK. I attempted to run the example downloaded from the webpage (the previous version), but I continued to encounter issues. Iterations occurred, but the expected outputs weren't generated.
Today, I tried using the example called by PMtutorial(), and everything seemed to work properly. Iterations occurred in R, and no errors appeared.
I don't fully understand it because I am used to previous commands. Tomorrow Dr. Otálvaro is going to check with me that everything  is functioning correctly.
It is a relief that Pmetrics is running again, thank you very very much for help.

**Comment by Siel** [comment]
Hi @CeciKtz I just saw your message on Twitter. If you're ok with that please send me an email juliandavid347@gmail.com. I think it might be faster if you share your screen so I can see what it is happening

**Comment by mnneely** [comment]
Hi @CeciKtz, if you wish, you can stay on the dev branch, or you can return to the main installation with `remotes::install_github("LAPKB/Pmetrics")`.  We've fixed the 64-bit detection in the main branch. You won't have to use PmetricsData for now, although eventually we will be coming back to that.

**Comment by CeciKtz** [comment]
Hi! Just to close this discussion and in case somebody might need it: the last issue was that Pmetrics didn't work with  legacy commands and it couldn't generate the Report. Julian pointed out that I needed to install 'pandoc' package. 
In R console: install.packages("pandoc") 
In the Terminal with this command: brew install pandoc (I have already installed brew at that point).
Afterwards Pmetrics works excellent.
I wanted to thank each one for the immense support and concern.

**Comment by alee8868** [comment]
Hi,

I am also trying to troubleshoot what seems a similar issue here. I also have entered PMbuild() after installing Pmetrics, but it cannot find gfortran or brew, nor install them. I believe I did manually install gfortran but I continue to run into this issue. I also deleted and re-installed Rstudio, and just updated my MacOS to Sequoia. 

This is the error I find on Rconsole:

> PMbuild()
Pmetrics cannot find required compiled binary files.
Pmetrics cannot detect gfortran and will attempt to download and install all components.
Do you agree? (Y/N)Y
Checking for Brew 
Pmetrics did not find the Homebrew Package Manager and will attempt to download and install it.
Do you wish to proceed? (Y/N)Y
A new terminal will open and ask for permission to install.  Enter your computer password.
Press any key when Homebrew installation is complete.

Pmetrics did not find GCC and will install it.
sh: brew: command not found
sh: brew: command not found
ERROR: Pmetrics did not install gfortan automatically.
Please install gfortran manually and then run PMbuild().
Go to http://www.lapk.org/Pmetrics_install.php for help.
ERROR: Pmetrics did not install gfortran automatically.
Please install gfortran manually and then run PMbuild().
Go to http://www.lapk.org/Pmetrics_install.php for help.
Warning messages:
1: In system(paste0(brew_path, " ls --versions gcc")) :
  error in running command
2: In system(paste0(brew_path, " install gcc")) : error in running command
> 

The terminal gave me this when it attempted to install brew and gfortran
Last login: Thu Nov 21 09:17:05 on ttys000
/Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/library/Pmetrics/mac/install_homebrew.sh ; exit;
adamlee@MacBook-Air-5 ~ % /Library/Frameworks/R.framework/Versions/4.4-x86_64/Resources/library/Pmetrics/mac/install_homebrew.sh ; exit;
curl: (56) The requested URL returned error: 404

Saving session...
...copying shared history...
...saving history...truncating history files...
...completed.

[Process completed]

Anyone know how to proceed forward? Thanks!
